consensus/src/liveness/leader_reputation.rs

```diff
@@ -25,70 +25,108 @@ pub trait MetadataBackend: Send + Sync {
         )?;
+
+        let max_returned_version = events.first().map_or(0, |first| first.transaction_version);
+
+        let new_block_events: Vec<NewBlockEvent> = itertools::process_results(
+            events
+                .into_iter()
+                .map(|event| bcs::from_bytes::<NewBlockEvent>(event.event.event_data())),
+            |iter| iter.filter(|e| e.epoch() == self.epoch).collect(),
+        )?;
+
+        let hit_end = new_block_events.len() < limit;
+
+        let result = (
+            new_block_events,
+            std::cmp::max(lastest_db_version, max_returned_version),
+            hit_end,
+        );
+        *self.db_result.lock() = result.clone();
+        Ok(result)
+    }
+
+    fn get_from_db_result(
+        &self,
+        target_round: Round,
+        events: &Vec<NewBlockEvent>,
+        hit_end: bool,
+    ) -> Vec<NewBlockEvent> {
         let mut result = vec![];
         for event in events {
-            let e = bcs::from_bytes::<NewBlockEvent>(event.event.event_data())?;
-            if e.epoch() == self.epoch
-                && e.round() <= target_round
-                && result.len() < self.window_size
-            {
-                result.push((event.transaction_version, e));
+            if event.round() <= target_round && result.len() < self.window_size {
+                result.push(event.clone());
             }
         }
-        *self.window.lock() = result;
-        Ok(())
+
+        if result.len() < self.window_size && !hit_end {
+            error!("We are not fetching far enough in history, we filtered from {} to {}, but asked for {}", events.len(), result.len(), self.window_size);
+        }
+        result
     }
 }
 
 impl MetadataBackend for AptosDBBackend {
     // assume the target_round only increases
     fn get_block_metadata(&self, target_round: Round) -> Vec<NewBlockEvent> {
-        let (known_version, known_round) = self
-            .window
-            .lock()
-            .first()
-            .map(|(v, e)| (*v, e.round()))
-            .unwrap_or((0, 0));
-        if !(known_round == target_round
-            || known_version == self.aptos_db.get_latest_version().unwrap_or(0))
-        {
-            if let Err(e) = self.refresh_window(target_round) {
-                error!(
-                    error = ?e, "[leader reputation] Fail to refresh window",
-                );
-                return vec![];
+        let db_result = self.db_result.lock();
+        let events = &db_result.0;
+        let version = db_result.1;
+        let hit_end = db_result.2;
+
+        let has_larger = events.first().map_or(false, |e| e.round() >= target_round);
+        let lastest_db_version = self.aptos_db.get_latest_version().unwrap_or(0);
+        // check if fresher data has potential to give us different result
+        if !has_larger && version < lastest_db_version {
+            // we need to drop the lock before refreshing.
+            drop(db_result);
```

Comment:
Hmm..dropping the lock here means we might have multiple concurrent `refresh_db_result` calls - semantically that should be fine but we unnecessarily make additional DB call. Does it make sense to drop the lock in `refresh_db_result` call instead?